### ISSUE TITLE
On Beautifing code/text will be duplicated

### ISSUE THREAD

**Issue by lampe** [Glavin001/atom-beautify#331]
Hey,
when i do the following in a file with 200+ lines:

- ctrl+a (to select everything)
- run beautify

now i have everything duplicated when i now hit:

- ctrl+z (undo the duplication)
- run beautify again
Then it wont copy it any more.

**Comment by Glavin001** [comment]
That's strange. And why are you using ctrl+a to select everything? By default it will use the entire file.

This sounds like a bug with the selection code.
- https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L36-L43
- https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L66
- https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L94-L107
- https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L132-L144

I have a few other issues to tackle before I can look into this further. I definitely will be taking a close look at this bug as soon as I can.

Is this a new bug you've been seeing with a recent update?

**Comment by lampe** [comment]
the select all thing is just a habit 
no its now a new thing i started somewhere about version 0.24

**Comment by Glavin001** [comment]
I am unable to reproduce, by selecting all text and then beautifying, it correctly performs this and does not duplicate the code.

Could you update and provide me with some more information for how to reproduce. Running command `Atom Beautify - Help Debug Editor` would help.

**Comment by lampe** [comment]
# Atom Beautify - Debugging information

The following debugging information was generated by `Atom Beautify` on `Wed May 20 2015 11:09:13 GMT+0200 (CEST)`.

---

**Platform**: linux

## Versions

**Atom Version**: 0.199.0

**Atom Beautify Version**: 0.27.1

## Original file to be beautified

**Original File Path**: `/home/micha/dev/pflanzmich/projektlager/meteor_prototype/client/views/account/account.js`

**Original File Grammar**: JavaScript

**Original File Contents**: 
```javascript
Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};
```

## Beautification options

**Editor Options**: 
Options from Atom Editor settings
```json
{
    "indent_size": 2,
    "indent_char": " ",
    "indent_with_tabs": false
}
```

**Config Options**: 
Options from Atom Beautify package settings
```json
{
    "cs": {
        "configPath": ""
    },
    "c": {
        "configPath": ""
    },
    "cpp": {
        "configPath": ""
    },
    "css": {
        "indent_size": 2,
        "indent_char": " ",
        "selector_separator_newline": false,
        "newline_between_rules": false,
        "preserve_newlines": false
    },
    "d": {
        "configPath": ""
    },
    "html": {
        "indent_inner_html": false,
        "indent_size": 2,
        "indent_char": " ",
        "brace_style": "collapse",
        "indent_scripts": "normal",
        "wrap_line_length": 250,
        "wrap_attributes": "auto",
        "wrap_attributes_indent_size": 2,
        "preserve_newlines": true,
        "max_preserve_newlines": 10,
        "unformatted": [
            "a",
            "sub",
            "sup",
            "b",
            "i",
            "u"
        ],
        "end_with_newline": false
    },
    "java": {
        "configPath": ""
    },
    "js": {
        "indent_size": 2,
        "indent_char": " ",
        "indent_level": 0,
        "indent_with_tabs": false,
        "preserve_newlines": true,
        "max_preserve_newlines": 10,
        "space_in_paren": false,
        "jslint_happy": false,
        "space_after_anon_function": false,
        "brace_style": "collapse",
        "break_chained_methods": false,
        "keep_array_indentation": false,
        "keep_function_indentation": false,
        "space_before_conditional": true,
        "eval_code": false,
        "unescape_strings": false,
        "wrap_line_length": 0,
        "end_with_newline": false
    },
    "objectivec": {
        "configPath": ""
    },
    "pawn": {
        "configPath": ""
    },
    "perl": {
        "perltidy_profile": ""
    },
    "php": {
        "fixers": "",
        "level": ""
    },
    "python": {
        "max_line_length": 79,
        "indent_size": 2,
        "ignore": [
            "E24"
        ]
    },
    "sql": {
        "indent_size": 2,
        "keywords": "upper",
        "identifiers": "lower"
    },
    "vala": {
        "configPath": ""
    }
}
```

**Home Options**: 
Options from `/home/micha/.jsbeautifyrc`
```json
{}
```

**EditorConfig Options**: 
Options from [EditorConfig](http://editorconfig.org/) file
```json
{
    "isFulfilled": false,
    "isRejected": false
}
```

**Project Options**: 
Options from `.jsbeautifyrc` files starting from directory `/home/micha/dev/pflanzmich/projektlager/meteor_prototype/client/views/account` and going up to root
```json
[
    {},
    {},
    {},
    {},
    {},
    {},
    {},
    {},
    {}
]
```

## Results

**Beautified File Contents**: 
```javascript
Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};
Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};
```

**Logs**: 
```
2015-05-20T09:09:13.572Z - info: [/home/micha/.atom/packages/atom-beautify/src/beautifiers/index.coffee] beautify Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};Router.route("accounts", {
  template: "accounts",
  layoutTemplate: "appWithoutHeader",
  loadingTemplate: "loading",
});

Session.setDefault("AccountsShowLogin", true);

Template.accounts.rendered = function AccountsRendered() {};

Template.accounts.helpers({
  showLogin: function AccountsShowLogin() {
    return Session.get("AccountsShowLogin");
  }
});

Template.accounts.validateEmail = function validateEmail(email) {
  var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
}; [ { indent_size: 2, indent_char: ' ', indent_with_tabs: false },
  { cs: { configPath: '' },
    c: { configPath: '' },
    cpp: { configPath: '' },
    css: 
     { indent_size: 2,
       indent_char: ' ',
       selector_separator_newline: false,
       newline_between_rules: false,
       preserve_newlines: false },
    d: { configPath: '' },
    html: 
     { indent_inner_html: false,
       indent_size: 2,
       indent_char: ' ',
       brace_style: 'collapse',
       indent_scripts: 'normal',
       wrap_line_length: 250,
       wrap_attributes: 'auto',
       wrap_attributes_indent_size: 2,
       preserve_newlines: true,
       max_preserve_newlines: 10,
       unformatted: [Object],
       end_with_newline: false },
    java: { configPath: '' },
    js: 
     { indent_size: 2,
       indent_char: ' ',
       indent_level: 0,
       indent_with_tabs: false,
       preserve_newlines: true,
       max_preserve_newlines: 10,
       space_in_paren: false,
       jslint_happy: false,
       space_after_anon_function: false,
       brace_style: 'collapse',
       break_chained_methods: false,
       keep_array_indentation: false,
       keep_function_indentation: false,
       space_before_conditional: true,
       eval_code: false,
       unescape_strings: false,
       wrap_line_length: 0,
       end_with_newline: false },
    objectivec: { configPath: '' },
    pawn: { configPath: '' },
    perl: { perltidy_profile: '' },
    php: { fixers: '', level: '' },
    python: { max_line_length: 79, indent_size: 2, ignore: [Object] },
    sql: { indent_size: 2, keywords: 'upper', identifiers: 'lower' },
    vala: { configPath: '' } },
  {},
  { _bitField: 536870912,
    _fulfillmentHandler0: undefined,
    _rejectionHandler0: 
     { _bitField: 0,
       _fulfillmentHandler0: undefined,
       _rejectionHandler0: undefined,
       _progressHandler0: undefined,
       _promise0: undefined,
       _receiver0: undefined,
       _settledValue: undefined },
    _progressHandler0: undefined,
    _promise0: undefined,
    _receiver0: undefined,
    _settledValue: undefined },
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {} ] JavaScript /home/micha/dev/pflanzmich/projektlager/meteor_prototype/client/views/account/account.js
2015-05-20T09:09:13.573Z - verbose: [/home/micha/.atom/packages/atom-beautify/src/beautifiers/index.coffee] JavaScript name=JavaScript, namespace=js, grammars=[JavaScript], extensions=[js], type=integer, default=2, minimum=0, description=Indentation size/length (Supported by JS Beautify, Pretty Diff), title=JavaScript - Indent size, beautifiers=[JS Beautify, Pretty Diff], type=string, default= , minimum=0, description=Indentation character (Supported by JS Beautify, Pretty Diff), title=JavaScript - Indent char, beautifiers=[JS Beautify, Pretty Diff], type=integer, default=0, description=Initial indentation level (Supported by JS Beautify), title=JavaScript - Indent level, beautifiers=[JS Beautify], type=boolean, default=false, description=Indentation uses tabs, overrides `Indent Size` and `Indent Char` (Supported by JS Beautify), title=JavaScript - Indent with tabs, beautifiers=[JS Beautify], type=boolean, default=true, description=Preserve line-breaks (Supported by JS Beautify, Pretty Diff), title=JavaScript - Preserve newlines, beautifiers=[JS Beautify, Pretty Diff], type=integer, default=10, description=Number of line-breaks to be preserved in one chunk (Supported by JS Beautify), title=JavaScript - Max preserve newlines, beautifiers=[JS Beautify], type=boolean, default=false, description=Add padding spaces within paren, ie. f( a, b ) (Supported by JS Beautify), title=JavaScript - Space in paren, beautifiers=[JS Beautify], type=boolean, default=false, description=Enable jslint-stricter mode (Supported by JS Beautify), title=JavaScript - Jslint happy, beautifiers=[JS Beautify], type=boolean, default=false, description=Add a space before an anonymous function's parens, ie. function () (Supported by JS Beautify), title=JavaScript - Space after anon function, beautifiers=[JS Beautify], type=string, default=collapse, enum=[collapse, expand, end-expand, none], description=[collapse|expand|end-expand|none] (Supported by JS Beautify), title=JavaScript - Brace style, beautifiers=[JS Beautify], type=boolean, default=false, description=Break chained method calls across subsequent lines (Supported by JS Beautify), title=JavaScript - Break chained methods, beautifiers=[JS Beautify], type=boolean, default=false, description=Preserve array indentation (Supported by JS Beautify), title=JavaScript - Keep array indentation, beautifiers=[JS Beautify], type=boolean, default=false, description= (Supported by JS Beautify), title=JavaScript - Keep function indentation, beautifiers=[JS Beautify], type=boolean, default=true, description= (Supported by JS Beautify), title=JavaScript - Space before conditional, beautifiers=[JS Beautify], type=boolean, default=false, description= (Supported by JS Beautify), title=JavaScript - Eval code, beautifiers=[JS Beautify], type=boolean, default=false, description=Decode printable characters encoded in xNN notation (Supported by JS Beautify), title=JavaScript - Unescape strings, beautifiers=[JS Beautify], type=integer, default=0, description=Wrap lines at next opportunity after N characters (Supported by JS Beautify), title=JavaScript - Wrap line length, beautifiers=[JS Beautify], type=boolean, default=false, description=End output with newline (Supported by JS Beautify), title=JavaScript - End with newline, beautifiers=[JS Beautify], beautifiers=[JS Beautify, Pretty Diff, Pretty Diff, Pretty Diff, Pretty Diff]
2015-05-20T09:09:13.574Z - verbose: [/home/micha/.atom/packages/atom-beautify/src/beautifiers/index.coffee] beautifier JS Beautify silly=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, debug=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, verbose=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, info=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, warn=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, error=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, onLogging=function (handler) {
        var subscription;
        subscription = emitter.on('logging', handler);
        return subscription;
      }, silly=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, debug=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, verbose=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, info=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, warn=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, error=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, onLogging=function (handler) {
        var subscription;
        subscription = emitter.on('logging', handler);
        return subscription;
      }, languages=[HTML, Handlebars, Mustache, Marko, JavaScript, JSON, CSS], silly=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, debug=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, verbose=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, info=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, warn=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, error=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, onLogging=function (handler) {
        var subscription;
        subscription = emitter.on('logging', handler);
        return subscription;
      }, silly=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, debug=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, verbose=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, info=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, warn=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, error=function (msg) {
      // build argument list (level, msg, ... [string interpolate], [{metadata}], [callback])
      var args = [level].concat(Array.prototype.slice.call(arguments));
      target.log.apply(target, args);
    }, onLogging=function (handler) {
        var subscription;
        subscription = emitter.on('logging', handler);
        return subscription;
      }, languages=[CSV, ERB, EJS, HTML, XML, Spacebars, JSX, JavaScript, CSS, SCSS, Sass, JSON, TSS, LESS]

```

**Comment by Glavin001** [comment]
Like I said, I have been unable to reproduce. It is especially difficult to narrow down a problem on my end when this issue only occurs *sometimes*. If other users are experiencing this then hopefully they let me know and we can all work together to figure out why this would be the case. 

The tests are all passing -- both Travis CI and including what you pasted above -- so we know it is not in the beautification (the beautifiers are not duplicating the code) and that leaves the code for replacing the selection with the beautified text (see https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L111-L116 ) which makes sense for where the problem is. 

I suppose it could make sense if somehow in the Atom lifecycle to deselect your selected text (selected all) and therefore when the [`editor.getSelectedBufferRange`](https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L112) is somehow the top or bottom and not actually selecting everything, such that when [`editor.setTextInBufferRange`](https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L116) is run that it inserts / replaces an empty space with the newly beautified text and keeps the old text. 
This makes sense, however it would be more so an Atom bug with `editor.getSelectedBufferRange` not returning the appropriate positioning, or user error (deselecting the selected text while beautification was in progress), or another package causing the text to be deselected.

In summary, considering the above, I am going to close this issue and will be labelling as `wontfix`.

Note: It may be worthwhile to move the [`editor.getSelectedBufferRange`](https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L112) call [up closer to the top of the `beautify` function](https://github.com/Glavin001/atom-beautify/blob/master/src/beautify.coffee#L45) and so even if the user accidently deselected the text while beautification was in progress, that it would not matter.

**I would accept a Pull Request for something like that.** Thanks.

**Comment by Glavin001** [comment]
Another user is experiencing this (/cc @edpirone ), so let's reopen this discussion and see where it leads. Hopefully it is something simple to fix.

@edpirone you said
> Going through it now to see if there is a tag left open or something that might be causing the error.

Is this an error in Atom Beautify code you found or are you saying that when beautifying code that has a syntax error it duplicates the code?
